### PR TITLE
Add `AsyncSeq` vs `TaskSeq` comparison readme

### PR DESCRIPTION
**Pull request by abelbraaksma**
As in the title. Considering the questions asked about this, a table is warranted.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -163,6 +164,33 @@ let feedFromTwitter user pwd = taskSeq {
...
+| **Conversion to other**    | `TaskSeq.toAsyncSeq`                                                              | `AsyncSeq.toAsyncEnum`                                               |
+| **Conversion from other**  | Implicit (yield!) or `TaskSeq.ofAsyncSeq`                                         | `AsyncSeq.ofAsyncEnum`                                               |
+| **Recursion in `yield!`**  | No (requires F# support, upcoming)                                                | Yes                                                                  |
+| **Based on F# concept of** | `task`                                                                            | `async`                                                              |
+| **`MoveNextAsync`**        | `ValueTask<bool>`                                                                 | `Async<'T option>`                                                   |
+| **`Current` internals**    | `ValueOption<'T>`                                                                 | `Option<'T>`                                                         |
```
Comment (gusty) [inline README.md:189]: I would group these 2 in a single MoveNext/Current row. I think it's clearer as in AsyncSeq both things are in a single type.

**Review by gusty — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -163,6 +164,33 @@ let feedFromTwitter user pwd = taskSeq {
 }
 ```
 
+## Choosing between `AsyncSeq` and `TaskSeq`
+
+The [`AsyncSeq`][11] and `TaskSeq` library both operate on asynchronous sequences, but there are a few fundamental differences, most notably that the former _does not_ implement `IAsyncEnumerable<'T>`, but has its own same-named, but differently behaving type. Another core difference is that `TaskSeq` uses `ValueTasks` for the asynchronous computations, and `AsyncSeq` uses F#'s `Async<'T>`.
```
Comment (bartelink) [inline README.md:169]:
```suggestion
The [`AsyncSeq`][11] and `TaskSeq` libraries both operate on asynchronous sequences, but there are a few fundamental differences. The most notable being that the former _does not_ implement `IAsyncEnumerable<'T>`, though it does have a type of that name with different semantics (not surprising; it predates the definition of the modern one). Another key difference is that `TaskSeq` uses `ValueTask`s for the asynchronous computations, whereas `AsyncSeq` uses F#'s `Async<'T>`.
```

File: README.md [hunk 3]
```diff
@@ -163,6 +164,33 @@ let feedFromTwitter user pwd = taskSeq {
...
+|----------------------------|---------------------------------------------------------------------------------------|----------------------------------------------------------------------|
+| **Frameworks**             | .NET 5.0+, NetStandard 2.1                                                        | .NET 5.0+, NetStandard 2.0 and 2.1, .NET Framework 4.6.1+            |
+| **Underlying type**        | `System.Collections.Generic.IAsyncEnumerable<'T>`                                 | Its own type, also called `IAsyncEnumerable<'T>`, but not compatible |
+| **Implementation**         | State machine (statically compiled)                                               | No state machine, continuation style                                 |
+| **Semantics**              | `seq`-like: on-demand                                                             | `seq`-like: on-demand                                                |
+| **Support `let!`**         | All `task`-like: `Async<'T>`, `ValueTask<'T>`, `Task<'T>` or any `GetAwaiter()`   | `Async<'T>` only                                                     |
```
Comment (bartelink) [inline README.md:179]:
```suggestion
| **Support `let!`**         | All `task`-like: `Async<'T>`, `Task<'T>`, `ValueTask<'T>` or any `GetAwaiter()`   | `Async<'T>` only                                                     |
```
(VT last to align with next line)

File: README.md [hunk 4]
```diff
@@ -163,6 +164,33 @@ let feedFromTwitter user pwd = taskSeq {
...
+| **Support `do!`**          | `Async<unit>`, `Task<unit>` and `Task`, `ValueTask<unit>` and `ValueTask`         | `Async<unit>` only                                                   |
```
Comment (bartelink) [inline README.md:180]:
```suggestion
| **Support `do!`**          | `Async<unit>`, `Task<unit>`, `Task`, `ValueTask<unit>`, `ValueTask`         | `Async<unit>` only                                                   |
```

File: README.md [hunk 5]
```diff
@@ -163,6 +164,33 @@ let feedFromTwitter user pwd = taskSeq {
...
+| **Support `yield!`**       | `IAsyncEnumerable<'T>`, `AsyncSeq`, any sequence                                  | `AsyncSeq`                                                           |
+| **Support `for`**          | `IAsyncEnumerable<'T>`, `AsyncSeq`, any sequence                                  | `AsyncSeq` any sequence                                              |
```
Comment (bartelink) [inline README.md:182]:
```suggestion
| **Support `for`**          | `IAsyncEnumerable<'T>`, `AsyncSeq`, any sequence                                  | `AsyncSeq`, 
 any sequence                                              |
```

File: README.md [hunk 6]
```diff
@@ -163,6 +164,33 @@ let feedFromTwitter user pwd = taskSeq {
...
+| **Behavior with `yield`**  | Zero allocations, no `Task` or even `ValueTask` created                           | Allocated, an F# `Async` wrapped in a singleton `AsyncSeq`           |
```
Comment (bartelink) [inline README.md:183]:
```suggestion
| **Behavior with `yield`**  | Zero allocations; no `Task` or even `ValueTask` created                           | Allocates an F# `Async` wrapped in a singleton `AsyncSeq`           |
```

File: README.md [hunk 7]
```diff
@@ -163,6 +164,33 @@ let feedFromTwitter user pwd = taskSeq {
...
+| **Conversion to other**    | `TaskSeq.toAsyncSeq`                                                              | `AsyncSeq.toAsyncEnum`                                               |
+| **Conversion from other**  | Implicit (yield!) or `TaskSeq.ofAsyncSeq`                                         | `AsyncSeq.ofAsyncEnum`                                               |
```
Comment (bartelink) [inline README.md:185]:
```suggestion
| **Conversion from other**  | Implicit (`yield!`) or `TaskSeq.ofAsyncSeq`                                         | `AsyncSeq.ofAsyncEnum`                                               |
```

File: README.md [hunk 8]
```diff
@@ -163,6 +164,33 @@ let feedFromTwitter user pwd = taskSeq {
...
+| **Recursion in `yield!`**  | No (requires F# support, upcoming)                                                | Yes                                                                  |
```
Comment (bartelink) [inline README.md:186]:
```suggestion
| **Recursion in `yield!`**  | **NO** (requires F# support, upcoming)                                                | Yes                                                                  |
```
capital O can be dropped, but good to make people think...

File: README.md [hunk 9]
```diff
@@ -163,6 +164,33 @@ let feedFromTwitter user pwd = taskSeq {
...
+| **Based on F# concept of** | `task`                                                                            | `async`                                                              |
```
Comment (bartelink) [inline README.md:187]: If this is going to be mentioned, it needs to be on line 2 or 3?

File: README.md [hunk 10]
```diff
@@ -163,6 +164,33 @@ let feedFromTwitter user pwd = taskSeq {
...
+| **`MoveNextAsync`**        | `ValueTask<bool>`                                                                 | `Async<'T option>`                                                   |
+| **`Current` internals**    | `ValueOption<'T>`                                                                 | `Option<'T>`                                                         |
+| **Cancellation**           | Implicit token governing iteration                                                | Implicit token passed to each subtask                                |
```
Comment (bartelink) [inline README.md:190]:
```suggestion
| **Cancellation**           | Implicit token governs iteration                                                | Implicit token flows to all subtasks per normal `Async` semantics                               |
```
Is "Implicit token governs iteration" -> "Implicit token governs iteration but requires explicit propagation of token to child Tasks"

File: README.md [hunk 11]
```diff
@@ -163,6 +164,33 @@ let feedFromTwitter user pwd = taskSeq {
...
+| **Performance**            | Very high, negligible allocations                                                 | Slower, more allocations, due to using `async`                       |
+| **Parallelism**            | Possible with ChildTask, support will follow                                      | Supported explicitly                                                 |
```
Comment (bartelink) [inline README.md:192]:
```suggestion
| **Parallelism**            | Possible with ChildTask; support will follow                                      | Extensive explicit helpers provided                                                 |
```

File: README.md [hunk 12]
```diff
@@ -163,6 +164,33 @@ let feedFromTwitter user pwd = taskSeq {
...
+
 ## Status & planning
 
 This project has stable features currently, but before we go full "version one", we'd like to complete the surface area. This section covers the status of that, with a full list of implmented functions below. Here's the short list:
```
Comment (bartelink) [inline README.md:196]:
```suggestion
This project has stable features currently, but before we go full "version one", we'd like to complete the surface area. This section covers the status of that, with a full list of implemented functions below. Here's the shortlist:
```

File: README.md (hunk 2 above)
Comment (abelbraaksma) [reply README.md]: Thanks! Done.

File: README.md [hunk 13]
```diff
@@ -163,6 +164,33 @@ let feedFromTwitter user pwd = taskSeq {
...
+
+The [`AsyncSeq`][11] and `TaskSeq` library both operate on asynchronous sequences, but there are a few fundamental differences, most notably that the former _does not_ implement `IAsyncEnumerable<'T>`, but has its own same-named, but differently behaving type. Another core difference is that `TaskSeq` uses `ValueTasks` for the asynchronous computations, and `AsyncSeq` uses F#'s `Async<'T>`.
+
+There are more differences:
+
+|                            | `TaskSeq`                                                                             | `AsyncSeq`                                                           |
+|----------------------------|---------------------------------------------------------------------------------------|----------------------------------------------------------------------|
+| **Frameworks**             | .NET 5.0+, NetStandard 2.1                                                        | .NET 5.0+, NetStandard 2.0 and 2.1, .NET Framework 4.6.1+            |
+| **Underlying type**        | `System.Collections.Generic.IAsyncEnumerable<'T>`                                 | Its own type, also called `IAsyncEnumerable<'T>`, but not compatible |
```
[the remark this answers is not in the document]
Comment (abelbraaksma) [reply README.md:176]: I'll keep both, I used your updated text.

File: README.md (hunk 3 above)
Comment (abelbraaksma) [reply README.md:179]: Tx, sharp eye!

File: README.md (hunk 4 above)
Comment (abelbraaksma) [reply README.md:180]: I like to keep this one, as it is deliberate (I had commas before but didn't like it), it signifies they're interchangeable semantically.

File: README.md (hunk 6 above)
Comment (abelbraaksma) [reply README.md:183]: Thanks!

File: README.md (hunk 8 above)
Comment (abelbraaksma) [reply README.md:186]: Good point

File: README.md (hunk 9 above)
Comment (abelbraaksma) [reply README.md:187]: Let's keep it separate for now.

File: README.md (hunk 11 above)
Comment (abelbraaksma) [reply README.md:192]: I'll take the semicolon, keep the original wording. People can always go to the docs to find out.

File: README.md (hunk 12 above)
Comment (abelbraaksma) [reply README.md:196]: Oops! lol. Tx.

File: README.md (hunk 1 above)
Comment (abelbraaksma) [reply README.md:189]: Tx. I'll probably just drop the second line altogether. Primarily, I don't want the left-most column to be too wordy and I can't think of a good way to explain both. Curious people can always look it up.

Comment (abelbraaksma) [reply README.md]: Done, tx.

File: README.md (hunk 10 above)
Comment (abelbraaksma) [reply README.md]: applied, tx

**Comment by abelbraaksma** [comment]
Thanks 👍 @gusty and @bartelink for the extensive review. I essentially applied all your suggestions (perhaps with very minor adjustments in the end). Squashed now (it got a bit messy in the end).
